### PR TITLE
Problem: No enclave-to-enclave mutually attested TLS transmission of tx + key payloads

### LINKED ISSUE #1549: Problem: no enclave-to-enclave mutually attested TLS transmission of tx+key payloads
https://github.com/crypto-com/chain-docs/blob/master/plan.md#datakey-retrieval-enclave-to-enclave-communication
so one will need to write protocol to choose whether to fetch key or txids.
for server-client certs, IAS etc., it can probably reuse or adapt some of the code written for new tx query?

what txids to fetch or what enclave IP to connect is out of the scope of this issue -- they could be provided as command line arguments (or have some other way of inputting them)

### PR DESCRIPTION
**Pull request by devashishdxt**
Solution: Created `tdbe` server implementation with one request types in protocol. Fixes #1549.

### CODE REVIEW

**Comment by devashishdxt** [comment]
bors try

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs [hunk 1]
```diff
@@ -212,19 +216,25 @@ impl EnclaveCertVerifier {
                 return Err(EnclaveCertVerifierError::MeasurementMismatch);
             }
 
-            if let Some(ref mr_enclave) = enclave_info.mr_enclave {
-                if mr_enclave != &quote.report_body.measurement.mr_enclave {
-                    return Err(EnclaveCertVerifierError::MeasurementMismatch);
+            match (
```
Comment (tomtau) [inline chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs:219]: would be good to add a comment about the validation logic here (https://github.com/crypto-com/chain-docs/blob/master/docs/modules/tdbe.md#svn-verification--compilation-order )

File: chain-tx-enclave-next/tdbe/app-runner/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "tdb-app-runner"
+version = "0.5.0"
```
Comment (tomtau) [inline chain-tx-enclave-next/tdbe/app-runner/Cargo.toml:3]:
```suggestion
version = "0.6.0"
```

File: chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "tdb-enclave-app"
+version = "0.5.0"
```
Comment (tomtau) [inline chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml:3]:
```suggestion
version = "0.6.0"
```

**Comment by devashishdxt** [comment]
bors try

File: chain-tx-enclave-next/mls/src/group.rs [hunk 4]
```diff
@@ -853,8 +857,11 @@ mod test {
             let now = 1590490084;
             let t = kp.verify(&*ENCLAVE_CERT_VERIFIER, now).unwrap();
 
+            let mut public_key = [0u8; 65];
+            public_key.copy_from_slice(certificate);
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/group.rs:861]: certificate here is the full X509 payload, not just the pubkey, so this is incorrect

File: docker/build.sh [hunk 5]
```diff
@@ -52,6 +52,12 @@ if [ $BUILD_MODE == "sgx" ]; then
     ftxsgx-elf2sgxs $EDP_TARGET_DIR/tx-query2-enclave-app --heap-size 0x2000000 --stack-size 0x80000 --threads 6 $EDP_ARGS
     sgxs-sign --key chain-tx-enclave/tx-validation/enclave/Enclave_private.pem $EDP_TARGET_DIR/tx-query2-enclave-app.sgxs $EDP_TARGET_DIR/tx-query2-enclave-app.sig -d --xfrm 7/0 --isvprodid 0 --isvsvn 0
 
+    # tdbe
+    cargo build --package tdb-app-runner
+    RUSTFLAGS="-Ctarget-feature=+aes,+sse2,+sse4.1,+ssse3,+pclmul" cargo build --target x86_64-fortanix-unknown-sgx --package tdb-enclave-app
+    ftxsgx-elf2sgxs $EDP_TARGET_DIR/tdb-enclave-app --heap-size 0x2000000 --stack-size 0x80000 --threads 6 $EDP_ARGS
```
Comment (tomtau) [inline docker/build.sh:58]: are such large stack sizes needed?
in default Enclave.config.xml, it's 0x40000

File: enclave-protocol/src/codec.rs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        to_send.append(&mut bytes);
+
+        writer.write(&to_send)
+    }
+}
+
+impl<T: Decode> StreamRead for T {
+    fn read_from<R: Read>(mut reader: R) -> io::Result<Self> {
+        let mut size = [0u8; 8];
+        reader.read_exact(&mut size)?;
```
Comment (tomtau) [inline enclave-protocol/src/codec.rs:45]: would there be any use case for transferring 18.446744073709553 _Exabytes_ ? or why u64 lengths without any sanity checks?

File: enclave-protocol/src/tdbe_protocol.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    GetTransactionsWithOutputs {
+        /// Transaction IDs for which to fetch all the transactions
+        transaction_ids: Cow<'a, [TxId]>,
+    },
+    /// Fetches keypackage for current node
+    GetKeyPackage,
+}
+
+#[derive(Encode, Decode)]
+pub enum TdbeResponse<'a> {
```
Comment (tomtau) [inline enclave-protocol/src/tdbe_protocol.rs:19]: probably good to make sure this is enclave-to-enclave response only (`TxWithOutputs` can't leave the enclave in cleartext)

File: enclave-protocol/src/tdbe_protocol.rs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    /// Contains all the requested transactions with outputs
+    GetTransactionsWithOutputs {
+        /// Requested transactions
+        transactions: Vec<TxWithOutputs>,
+    },
+    /// Contains keypackage for current node
+    GetKeyPackage {
+        /// Raw keypackage data
+        key_package: Vec<u8>, // TODO: Concrete `KeyPackage` type?
+    },
```
Comment (tomtau) [inline enclave-protocol/src/tdbe_protocol.rs:29]: this won't be in the necessary in the enclave-to-enclave response

File: enclave-protocol/src/tdbe_protocol.rs [hunk 9]
```diff
@@ -0,0 +1,35 @@
+use std::borrow::Cow;
+
+use parity_scale_codec::{Decode, Encode};
+
+use chain_core::tx::{data::TxId, TxWithOutputs};
+
```
Comment (tomtau) [inline enclave-protocol/src/tdbe_protocol.rs:6]: may be good to create the response data type for enclave->untrusted/runner
where it'll return the *sealed* transactions or the node's keypackage

File: chain-tx-enclave-next/mls/src/group.rs (hunk 4 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/mls/src/group.rs:861]: I didn't change any logic here. Only changed this from `Vec<u8>` to `[u8; 65]`. @yihuang Any comments?

File: docker/build.sh (hunk 5 above)
Comment (devashishdxt) [reply docker/build.sh:58]: This is what we use for `tx-query`. We can try and optimise these sizes later.

File: chain-tx-enclave-next/tdbe/app-runner/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/app-runner/Cargo.toml]: Done

File: chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml]: Done

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs [hunk 10]
```diff
@@ -111,23 +111,27 @@ impl EnclaveCertVerifier {
...
         let attestation_report_oid = Oid::from(OID_EXTENSION_ATTESTATION_REPORT)
             .expect("Unable to parse attestation report OID");
-        let public_key = certificate
-            .tbs_certificate
-            .subject_pki
-            .subject_public_key
-            .data;
+
+        let mut public_key = [0; 65];
+        // This will not panic because public key in certificate should be in uncompressed raw format (65 bytes)
```
[the remark this answers is not in the document]
Comment (devashishdxt) [reply chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs]: Added a length check.

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs:230]: Done

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (devashishdxt) [reply chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs]: https://github.com/crypto-com/chain/pull/1940#discussion_r454108326

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs (hunk 1 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs:230]: https://github.com/crypto-com/chain/pull/1940#discussion_r454108375

File: chain-tx-enclave-next/tdbe/app-runner/Cargo.toml (hunk 2 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/app-runner/Cargo.toml]: Done

File: chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml (hunk 3 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/enclave-app/Cargo.toml]: Done

File: enclave-protocol/src/codec.rs (hunk 6 above)
Comment (devashishdxt) [reply enclave-protocol/src/codec.rs]: Changed from `u64` to `u32`.

File: docker/build.sh (hunk 5 above)
Comment (tomtau) [reply docker/build.sh:58]: but is that necessary? it'll crash with 0x40000?

Comment (devashishdxt) [reply docker/build.sh:58]: I'll test if it works with smaller heap size.

File: chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs [hunk 11]
```diff
@@ -111,23 +111,37 @@ impl EnclaveCertVerifier {
...
             .subject_public_key
-            .data;
+            .data
+            .len()
+            != 65
+        {
+            return Err(EnclaveCertVerifierError::PublicKeyMismatch);
+        }
+
+        let mut public_key = [0; 65];
```
Comment (leejw51crypto) [inline chain-tx-enclave-next/enclave-ra/ra-client/src/verifier.rs:126]: how about using const?

File: chain-tx-enclave-next/tdbe/app-runner/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        Box::pin(connect_stream_inner(self, addr))
+    }
+
+    fn bind_stream<'future>(
+        &'future self,
+        addr: &'future str,
+        _local_addr: Option<&'future mut String>,
+    ) -> Pin<Box<dyn Future<Output = io::Result<Option<Box<dyn AsyncListener>>>> + 'future>> {
```
Comment (leejw51crypto) [inline chain-tx-enclave-next/tdbe/app-runner/src/main.rs:123]: how about using `type` to simplify this line?

File: enclave-protocol/src/tdbe_protocol.rs (hunk 9 above)
Comment (devashishdxt) [reply enclave-protocol/src/tdbe_protocol.rs:6]: Split the protocol in trusted/untrusted parts

File: enclave-protocol/src/tdbe_protocol.rs (hunk 8 above)
Comment (devashishdxt) [reply enclave-protocol/src/tdbe_protocol.rs]: https://github.com/crypto-com/chain/pull/1940#discussion_r454789009

File: enclave-protocol/src/tdbe_protocol.rs (hunk 7 above)
Comment (devashishdxt) [reply enclave-protocol/src/tdbe_protocol.rs]: https://github.com/crypto-com/chain/pull/1940#discussion_r454789009

File: chain-tx-enclave-next/tdbe/app-runner/src/main.rs (hunk 12 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/app-runner/src/main.rs:123]: We don't use this type anywhere else. Its just the type returned by this trait: https://docs.rs/enclave-runner/0.4.0/enclave_runner/usercalls/trait.UsercallExtension.html

**Comment by devashishdxt** [comment]
bors try

**Comment by devashishdxt** [comment]
bors try

**Comment by devashishdxt** [comment]
bors try

**Comment by tomtau** [comment]
bors try-

**Comment by tomtau** [comment]
bors try

**Comment by devashishdxt** [comment]
bors try

File: chain-tx-enclave-next/tdbe/enclave-app/src/sgx_module.rs [hunk 13]
```diff
@@ -0,0 +1,262 @@
+mod handler;
+
+pub use rs_libc::alloc::*;
```
Comment (tomtau) [inline chain-tx-enclave-next/tdbe/enclave-app/src/sgx_module.rs:3]: I haven't tried, but wonder if this is still necessary? With the upstream(-ish) rust-secp256k1, allocations+abort... are done in Rust, so no need for special patches and libc

Comment (tomtau) [reply chain-tx-enclave-next/tdbe/enclave-app/src/sgx_module.rs:3]: Does ring need libc allocs?

File: integration-tests/bot/chainbot.py [hunk 14]
```diff
@@ -236,6 +238,8 @@ def programs(node, app_hash, root_path, cfg):
     if not cfg.get('mock_mode'):
         commands += [
                 ('tx-query', f"tx-query2-app-runner --enclave-path 'tx-query2-enclave-app.sgxs' --address 127.0.0.1:{tx_query_port} --zmq-conn-str tcp://127.0.0.1:{tx_validation_port} --sp-address 127.0.0.1:8989", dict(def_env)),
+                ('tdbe-1', f"tdb-app-runner --enclave-path 'tdb-enclave-app.sgxs' --address 127.0.0.1:{tdbe_1_port} --zmq-conn-str tcp://127.0.0.1:{tx_validation_port} --sp-address 127.0.0.1:8989", dict(def_env)),
+                ('tdbe-2', f"tdb-app-runner --enclave-path 'tdb-enclave-app.sgxs' --address 127.0.0.1:{tdbe_2_port} --zmq-conn-str tcp://127.0.0.1:{tx_validation_port} --sp-address 127.0.0.1:8989 --tdbe-address 127.0.0.1:{tdbe_1_port} --tdbe-dns-name localhost --txids 37d7eaa7987ab1d6c9d7d4ed4398cb3f326b02d35e2f8e9d8d824fa676fa3ce5", dict(def_env)),
```
Comment (tomtau) [inline integration-tests/bot/chainbot.py:242]: I'm not sure if all tests will have this tx `37d7eaa7987ab1d6c9d7d4ed4398cb3f326b02d35e2f8e9d8d824fa676fa3ce5` -- maybe there could just be a separate test which will launch current tdbe code, or leave it for now?

**Comment by devashishdxt** [comment]
bors try

File: integration-tests/bot/chainbot.py (hunk 14 above)
Comment (devashishdxt) [reply integration-tests/bot/chainbot.py:242]: We can change the tests in a different PR. For now, this transaction seems to work.

File: chain-tx-enclave-next/tdbe/enclave-app/src/sgx_module.rs (hunk 13 above)
Comment (devashishdxt) [reply chain-tx-enclave-next/tdbe/enclave-app/src/sgx_module.rs:3]: I'm investigating this. Will create a new PR if it isn't necessary anymore.

**Pull request merged**